experiments_main.cpp

```diff
@@ -107,119 +111,114 @@ DEFINE_uint32(num_threads,
-                    parse_hash_algorithm(FLAGS_hash_alg)),
-          omin_hash(int_pow<uint32_t>(FLAGS_alphabet_size, FLAGS_kmer_size),
-                    FLAGS_embed_dim,
-                    FLAGS_max_len,
-                    FLAGS_tuple_length,
-                    parse_hash_algorithm(FLAGS_hash_alg)),
-          tensor_sketch(FLAGS_alphabet_size, FLAGS_embed_dim, FLAGS_tuple_length),
-          tensor_slide(FLAGS_alphabet_size,
-                       ceil(sqrt(FLAGS_embed_dim)),
-                       FLAGS_tuple_length,
-                       FLAGS_window_size,
-                       FLAGS_stride),
-          l1SketchBin32(FLAGS_embed_dim, ceil(sqrt(FLAGS_embed_dim)), FLAGS_seq_len),
-          l1Sketch(FLAGS_embed_dim, ceil(sqrt(FLAGS_embed_dim)), FLAGS_seq_len) {}
+template <class char_type, class kmer_type, class embed_type, class... SketchAlgorithms>
+class ExperimentRunner {
+    static constexpr size_t NumAlgorithms = sizeof...(SketchAlgorithms);
+    std::tuple<SketchAlgorithms...> algorithms_;
 
-    void run() {
-        std::cout << "Generating sequences ..." << std::flush;
-        generate_sequences();
-        std::cout << "\nComputing sketches ... " << std::flush;
-        compute_sketches();
-        std::cout << "\nTransform sketches ... " << std::flush;
-        transform_sketches();
-        std::cout << "\nComputing distances ... " << std::flush;
-        compute_pairwise_dists();
-        std::cout << "\nComputing Spearman correlation ... \n" << std::flush;
-        print_summary();
-        std::cout << "Writing output to ... " << FLAGS_o << std::endl;
-        save_output();
-    }
+    Vec2D<char_type> seqs;
+    std::vector<std::pair<uint32_t, uint32_t>> ingroup_pairs;
 
-    void generate_sequences() {
-        ts::SeqGen seq_gen(FLAGS_alphabet_size, FLAGS_fix_len, FLAGS_num_seqs, FLAGS_seq_len,
-                           FLAGS_group_size, FLAGS_max_mutation_rate, FLAGS_min_mutation_rate,
-                           FLAGS_phylogeny_shape);
+    std::vector<embed_type> edit_dists;
+    using Distances
+            = std::tuple<std::conditional_t<true, std::vector<double>, SketchAlgorithms>...>;
+    Distances dists;
 
-        seqs = seq_gen.generate_seqs<char_type>();
-        seq_gen.ingroup_pairs(ingroup_pairs);
+  public:
+    ExperimentRunner() = delete;
+    ExperimentRunner(ExperimentRunner &) = delete;
 
+    explicit ExperimentRunner(SketchAlgorithms... algorithms) : algorithms_(algorithms...) {}
 
-        size_t num_seqs = seqs.size();
-        wmh_sketch.resize(num_seqs);
-        mh_sketch.resize(num_seqs);
-        omh_sketch.resize(num_seqs);
-        ts_sketch.resize(num_seqs);
-        tss_sketch.resize(num_seqs);
-        tss_sketch_flat.resize(num_seqs);
-        tss_sketch_binary.resize(num_seqs);
-    }
 
-    void compute_sketches() {
+    // Return the Spearman coefficient.
+    template <class SketchAlgorithm>
+    double run_sketch_algorithm(SketchAlgorithm &algorithm) {
```

Comment:
Sadly it can't, because the `compute` functions aren't always const (the minhash variants update the hash maps on the fly...)

I'll make it a pointer.